packages/firestore/src/core/firestore_client.ts

```diff
@@ -357,6 +367,57 @@ export class FirestoreClient {
     });
   }
 
+  getDocumentFromLocalCache(
+    doc: DocumentReference,
+    observer: Observer<DocumentSnapshot>
+  ): void {
+    this.asyncQueue.schedule(() => {
+      return this.localStore.readDocument(doc._key).then(maybeDoc => {
+        if (maybeDoc instanceof Document) {
+          observer.next(
+            new DocumentSnapshot(doc.firestore, doc._key, maybeDoc, true)
+          );
+        } else {
+          observer.error(
+            new Error(
+              'Failed to get document from cache. (However, this document may ' +
+                "exist on the server. Run again without setting 'source' in " +
+                'the GetOptions to attempt to retrieve the document from the ' +
+                'server.)'
+            )
+          );
+        }
+      });
+    });
+  }
+
+  getDocumentsFromLocalCache(
+    query: Query,
+    observer: Observer<QuerySnapshot>
+  ): void {
+    this.asyncQueue.schedule(() => {
+      return this.localStore.executeQuery(query._query).then(docs => {
+        const remoteKeys: DocumentKeySet = documentKeySet();
+
+        const view = new View(query._query, remoteKeys);
+        const viewDocChanges: ViewDocumentChanges = view.computeDocChanges(
+          docs
+        );
+        const viewChange: ViewChange = view.applyChanges(viewDocChanges);
+        assert(
+          viewChange.limboChanges.length === 0,
+          'View returned limbo documents during local-only query execution.'
+        );
```

Comment:
So I was actually surprised that this assert held because remoteKeys is empty, and so all the documents should be in limbo (they're in the local cache but not in the remote results). But it turns out we don't calculate limbo documents until you call applyChanges() with a TargetChange.

So this assert _holds_ but not for the reason I would have expected. And if there _were_ limbo documents, that wouldn't be a problem for us.  So I'd argue this assert is confusing and not protecting us from anything, so I'd just remove it.